### ISSUE TITLE
Softmax/logsoftmax use more memory for backward than needed

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#64000]
## 🚀 Feature
Backward formulas for softmax/logsoftmax reference `self`, even though `self` is not required for computation, e.g. 
```
- name: _log_softmax(Tensor self, int dim, bool half_to_float) -> Tensor
  self: _log_softmax_backward_data(grad, result, dim, self)
 ```
Therefore, we retain reference to `self` and use more memory than needed. We should try to avoid that. 
Related: #63027 (more memory than needed used for ReLU), with the same pitch and motivation. 

The explanation for adding `self` to backward/double backward formulas is in #6786, "in softmax_backward, self argument is added just so that gradient of self can be computed in double backward. Self tensor itself is not necessary neither for softmax_backward, nor for softmax_double_backward." There was no solution proposed in #6786. 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by heitorschueroff** [comment]
> Just leaving a thought here: I see that all these kernels are implemented manually on CPU and CUDA rather than via the more obvious `logsumexp`. I gather this is for efficiency to avoid calling 3 kernels, but is it that big a gain? (asking, I don't know) I remember @peterbell10 mentioning how nice is the algorithm that we have for `sum`, and how much more accurate it can be than just naïve accumulation. Would that apply here or would performance be more important for these specific functions?

I think it's worth at least benchmarking a composite implementation. Going forward, we should create an easy to use API for chaining transformation - reduction - projection so that we can have custom kernels for these operators without having to write a lot of code.

**Comment by blefaudeux** [comment]
> > Just leaving a thought here: I see that all these kernels are implemented manually on CPU and CUDA rather than via the more obvious `logsumexp`. I gather this is for efficiency to avoid calling 3 kernels, but is it that big a gain? (asking, I don't know) I remember @peterbell10 mentioning how nice is the algorithm that we have for `sum`, and how much more accurate it can be than just naïve accumulation. Would that apply here or would performance be more important for these specific functions?
> 
> I think it's worth at least benchmarking a composite implementation. Going forward, we should create an easy to use API for chaining transformation - reduction - projection so that we can have custom kernels for these operators without having to write a lot of code.

what's the backend of choice for that ? I'm testing out exactly this idea using Triton actually, it's not clear to me how you can do that without Jit (except if you're willing to support a lot of automatically generated kernels)

**Comment by ngimel** [comment]
Triton is not doing a composite operation, the main reason for triton excellent perf is that it combines reduction and transformation steps keeping the values in shared memory. 
For large sizes, where you can't keep the values in shared memory and use Triton's strategy, composite operation would have close performance to what our softmax kernel has now (given that reduction is implemented efficiently, that's an "if").
However, this is orthogonal to initial discussion of changing gradient formulas. We likely will still want softmax as a prim operation (not a composite) and thus will need efficient gradient formulas for it.

**Comment by blefaudeux** [comment]
@ngimel I was reacting to the proposal above, as I understood it (having a single kernel for chained transformations), I was curious of the pytorch-ic approach to that to make sure that I was not missing anything there. [fb only] I'm testing this approach with Triton for linear layers in https://github.com/fairinternal/xformers/pull/219 (not softmax related)
